In [1]:
import numpy as np
from scipy import stats
import pandas as pd

np.random.seed(5)

daily_temps = np.random.normal(
    loc=23,
    scale=2.5,
    size=30
)

mean = daily_temps.mean()
s = daily_temps.std(ddof=1)

print("Варіант 5 — Дніпро")
print("Денні температури:")
print(daily_temps)

print("\nСереднє:", mean)
print("Вибіркове стандартне відхилення:", s)

Варіант 5 — Дніпро
Денні температури:
[24.10306872 22.17282462 29.07692797 22.36976968 23.2740246  26.95620279
 20.72691899 21.52090836 23.46900806 22.17532511 20.01808847 22.48780872
 22.10292763 24.50867901 18.83802868 21.24955241 25.87847752 27.64332752
 19.22205111 24.61211878 20.54848029 20.85786711 20.82030204 21.94373018
 25.49109957 24.78105318 23.14786061 22.0917228  23.00822211 22.73517389]

Середнє: 22.92771835022647
Вибіркове стандартне відхилення: 2.4151177465764073


Для варіанта 5 (Дніпро) було згенеровано вибірку з 30 денних
температур.

Середня температура вибірки становить приблизно 22.93 °C,
а вибіркове стандартне відхилення — 2.42 °C.

Середнє близьке до табличного липневого значення 23 °C.

In [2]:
n = len(daily_temps)

se = s / np.sqrt(n)

ci_mean = stats.t.interval(
    0.95,
    df=n - 1,
    loc=mean,
    scale=se
)

print("n =", n)
print("Стандартна похибка:", se)
print("95% довірчий інтервал:", ci_mean)

n = 30
Стандартна похибка: 0.44093815627698113
95% довірчий інтервал: (np.float64(22.025898562661446), np.float64(23.829538137791495))


95% довірчий інтервал для середньої температури становить
приблизно [22.03 °C; 23.83 °C].

Тут використовується t-розподіл, а не z-розподіл, оскільки
справжнє стандартне відхилення генеральної сукупності σ
нам невідоме.

Ми маємо лише вибіркове стандартне відхилення s, обчислене
за 30 спостереженнями. t-розподіл враховує додаткову
невизначеність, пов'язану з оцінюванням σ за вибіркою.

In [3]:
p_hat = (daily_temps > 23).mean()

np_hat = n * p_hat
n1p_hat = n * (1 - p_hat)

se_p = np.sqrt(
    p_hat * (1 - p_hat) / n
)

ci_prop = stats.norm.interval(
    0.95,
    loc=p_hat,
    scale=se_p
)

print("Частка днів вище 23°C:", p_hat)
print("n * p_hat =", np_hat)
print("n * (1 - p_hat) =", n1p_hat)
print("95% довірчий інтервал:", ci_prop)

Частка днів вище 23°C: 0.43333333333333335
n * p_hat = 13.0
n * (1 - p_hat) = 17.0
95% довірчий інтервал: (np.float64(0.2560114457146738), np.float64(0.6106552209519929))


Частка днів із температурою вище 23 °C становить
0.4333, або приблизно 43.33%.

Перевірка нормального наближення:

n·p̂ = 13
n·(1-p̂) = 17

Обидва значення більші за 5, тому нормальне наближення
для побудови довірчого інтервалу є прийнятним.

95% довірчий інтервал для частки становить приблизно
[0.256; 0.611], або [25.60%; 61.07%].

Теоретичне значення 0.5 входить до отриманого інтервалу,
що узгоджується з тим, що вибірка була згенерована
симетрично навколо температури 23 °C.

In [4]:
confidence_levels = [0.90, 0.95, 0.99]

results = []

for level in confidence_levels:
    ci = stats.t.interval(
        level,
        df=n - 1,
        loc=mean,
        scale=se
    )

    width = ci[1] - ci[0]

    results.append({
        "Рівень довіри": level,
        "Нижня межа": ci[0],
        "Верхня межа": ci[1],
        "Ширина": width
    })

ci_table = pd.DataFrame(results)

ci_table

,Рівень довіри,Нижня межа,Верхня межа,Ширина
0,0.90,22.178508,23.676928,1.498420
1,0.95,22.025899,23.829538,1.803640
2,0.99,21.712323,24.143114,2.430791


Зі збільшенням рівня довіри ширина довірчого інтервалу
зростає.

90% → ширина ≈ 1.50 °C
95% → ширина ≈ 1.80 °C
99% → ширина ≈ 2.43 °C

Чим вищий рівень довіри ми хочемо отримати, тим ширшим
має бути інтервал, щоб процедура частіше накривала справжнє
значення параметра.

Щоб отримати вужчий інтервал при тому самому рівні довіри
99%, можна збільшити розмір вибірки n. Це зменшить
стандартну похибку s / √n.

Отже, за більшу точність потрібно "заплатити" більшою
кількістю спостережень.

Отриманий 95% довірчий інтервал становить приблизно
[22.03 °C; 23.83 °C].

Якби ми багато разів повторювали процедуру вибірки 30 днів
і щоразу будували довірчий інтервал таким самим способом,
то приблизно 95% таких інтервалів містили б справжню
середню температуру генеральної сукупності.

Приблизно 5% побудованих інтервалів її не містили б.

Контрольні питання

1. Яка різниця між точковою оцінкою і довірчим інтервалом?

Точкова оцінка — це одне число, отримане за вибіркою.
Наприклад, вибіркове середнє 22.93 °C є точковою оцінкою
справжньої середньої температури.

Одного числа недостатньо, оскільки воно не показує
невизначеність оцінювання.

Довірчий інтервал задає діапазон значень і показує
точність оцінювання параметра.


2. Чому використовується t-розподіл?

Для середнього використовується t-розподіл, оскільки
справжнє стандартне відхилення генеральної сукупності σ
невідоме.

Замість нього використовується вибіркове стандартне
відхилення s.

При невеликій вибірці це створює додаткову невизначеність,
яку враховує t-розподіл. Зі збільшенням n t-розподіл
наближається до нормального.


3. Що насправді означає 95% довіри?

95% довіри означає, що якби процедуру отримання вибірки
та побудови інтервалу повторювали багато разів, приблизно
95% отриманих інтервалів містили б справжній параметр.

Неправильно говорити, що після побудови конкретного
інтервалу існує 95% ймовірність знаходження параметра
в ньому.

У частотному підході справжній параметр є фіксованим,
а випадковими до отримання даних є межі інтервалу.


4. Які фактори визначають ширину довірчого інтервалу?

Основними факторами є:

1) Рівень довіри. Чим він вищий, тим ширший інтервал.

2) Розкид даних. Чим більше стандартне відхилення,
тим ширший довірчий інтервал.

3) Розмір вибірки n. Чим більше спостережень, тим менша
стандартна похибка і тим вужчий довірчий інтервал.

Таким чином, збільшення вибірки дозволяє підвищити
точність оцінювання без зменшення рівня довіри.